### ISSUE TITLE
no implicit conversion of nil into String (cache.rb line 182)

### ISSUE THREAD

**Issue by jgrannas** [carrierwaveuploader/carrierwave#2107]
I updated carrierwave/fog and I am using jquery file upload which worked great previously... but now after updating I can not seem to figure out what is going on with this error. I found a few others with similar issues but could not find a solution.

Using carrierwave 1.0.0
Using fog 1.38.0

`Started POST "/car_images" for 127.0.0.1 at 2017-01-30 14:22:21 -0500
Processing by CarImagesController#create as */*
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"W3bMZzVHrO1XcPedEUGyM8EgmkPxvOUu+nh8l8JlYol4AHQH2K474qxcfcyIvmnj36FpQnPkf2gdw0fboF+EfQ==", "car_image"=>{"car_id"=>"1", "file"=>[#<ActionDispatch::Http::UploadedFile:0x007fd065eadde0 @tempfile=#<Tempfile:/var/folders/88/4_kr10393hbctbvsk5nd0vf80000gn/T/RackMultipart20170130-827-1khwyf3.JPG>, @original_filename="DSC_6776.JPG", @content_type="image/jpeg", @headers="Content-Disposition: form-data; name=\"car_image[file][]\"; filename=\"DSC_6776.JPG\"\r\nContent-Type: image/jpeg\r\n">]}}
Completed 500 Internal Server Error in 4ms (ActiveRecord: 0.0ms)

TypeError (no implicit conversion of nil into String):
  carrierwave (1.0.0) lib/carrierwave/uploader/cache.rb:182:in `join'
  carrierwave (1.0.0) lib/carrierwave/uploader/cache.rb:182:in `workfile_path'
  carrierwave (1.0.0) lib/carrierwave/uploader/cache.rb:134:in `cache!'
  carrierwave (1.0.0) lib/carrierwave/mounter.rb:44:in `block in cache'
  carrierwave (1.0.0) lib/carrierwave/mounter.rb:42:in `map'
  carrierwave (1.0.0) lib/carrierwave/mounter.rb:42:in `cache'
  carrierwave (1.0.0) lib/carrierwave/mount.rb:146:in `file='
  carrierwave (1.0.0) lib/carrierwave/orm/activerecord.rb:70:in `file='
  activerecord (4.2.7.1) lib/active_record/attribute_assignment.rb:54:in `public_send'
  activerecord (4.2.7.1) lib/active_record/attribute_assignment.rb:54:in `_assign_attribute'
  activerecord (4.2.7.1) lib/active_record/attribute_assignment.rb:41:in `block in assign_attributes'
  actionpack (4.2.7.1) lib/action_controller/metal/strong_parameters.rb:185:in `each_pair'
  actionpack (4.2.7.1) lib/action_controller/metal/strong_parameters.rb:185:in `each_pair'
  activerecord (4.2.7.1) lib/active_record/attribute_assignment.rb:35:in `assign_attributes'
  activerecord (4.2.7.1) lib/active_record/core.rb:566:in `init_attributes'
  activerecord (4.2.7.1) lib/active_record/core.rb:281:in `initialize'
  actionpack (4.2.7.1) lib/action_dispatch/routing/url_for.rb:104:in `initialize'
  activerecord (4.2.7.1) lib/active_record/inheritance.rb:61:in `new'
  activerecord (4.2.7.1) lib/active_record/inheritance.rb:61:in `new'
  cancancan (1.15.0) lib/cancan/controller_resource.rb:88:in `build_resource'
  cancancan (1.15.0) lib/cancan/controller_resource.rb:69:in `load_resource_instance'
  cancancan (1.15.0) lib/cancan/controller_resource.rb:40:in `load_resource'
  cancancan (1.15.0) lib/cancan/controller_resource.rb:33:in `load_and_authorize_resource'
  cancancan (1.15.0) lib/cancan/controller_resource.rb:10:in `block in add_before_action'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:448:in `instance_exec'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:448:in `block in make_lambda'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:164:in `block in halting'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:504:in `block in call'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:504:in `each'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:504:in `call'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:92:in `__run_callbacks__'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:778:in `_run_process_action_callbacks'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.7.1) lib/abstract_controller/callbacks.rb:19:in `process_action'
  actionpack (4.2.7.1) lib/action_controller/metal/rescue.rb:29:in `process_action'
  actionpack (4.2.7.1) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
  activesupport (4.2.7.1) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (4.2.7.1) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
  activesupport (4.2.7.1) lib/active_support/notifications.rb:164:in `instrument'
  actionpack (4.2.7.1) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  actionpack (4.2.7.1) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
  activerecord (4.2.7.1) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  actionpack (4.2.7.1) lib/abstract_controller/base.rb:137:in `process'
  actionview (4.2.7.1) lib/action_view/rendering.rb:30:in `process'
  actionpack (4.2.7.1) lib/action_controller/metal.rb:196:in `dispatch'
  actionpack (4.2.7.1) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
  actionpack (4.2.7.1) lib/action_controller/metal.rb:237:in `block in action'
  actionpack (4.2.7.1) lib/action_dispatch/routing/route_set.rb:74:in `dispatch'
  actionpack (4.2.7.1) lib/action_dispatch/routing/route_set.rb:43:in `serve'
  actionpack (4.2.7.1) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.7.1) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.7.1) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.7.1) lib/action_dispatch/routing/route_set.rb:817:in `call'
  omniauth (1.3.2) lib/omniauth/strategy.rb:186:in `call!'
  omniauth (1.3.2) lib/omniauth/strategy.rb:164:in `call'
  jquery-fileupload-rails (0.4.7) lib/jquery/fileupload/rails/middleware.rb:14:in `_call'
  jquery-fileupload-rails (0.4.7) lib/jquery/fileupload/rails/middleware.rb:10:in `call'
  warden (1.2.7) lib/warden/manager.rb:36:in `block in call'
  warden (1.2.7) lib/warden/manager.rb:35:in `catch'
  warden (1.2.7) lib/warden/manager.rb:35:in `call'
  rack (1.6.5) lib/rack/etag.rb:24:in `call'
  rack (1.6.5) lib/rack/conditionalget.rb:38:in `call'
  rack (1.6.5) lib/rack/head.rb:13:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/flash.rb:260:in `call'
  rack (1.6.5) lib/rack/session/abstract/id.rb:225:in `context'
  rack (1.6.5) lib/rack/session/abstract/id.rb:220:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/cookies.rb:560:in `call'
  activerecord (4.2.7.1) lib/active_record/query_cache.rb:36:in `call'
  activerecord (4.2.7.1) lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'
  activerecord (4.2.7.1) lib/active_record/migration.rb:377:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:88:in `__run_callbacks__'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:778:in `_run_call_callbacks'
  activesupport (4.2.7.1) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/reloader.rb:73:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  railties (4.2.7.1) lib/rails/rack/logger.rb:38:in `call_app'
  railties (4.2.7.1) lib/rails/rack/logger.rb:20:in `block in call'
  activesupport (4.2.7.1) lib/active_support/tagged_logging.rb:68:in `block in tagged'
  activesupport (4.2.7.1) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (4.2.7.1) lib/active_support/tagged_logging.rb:68:in `tagged'
  railties (4.2.7.1) lib/rails/rack/logger.rb:20:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/request_id.rb:21:in `call'
  rack (1.6.5) lib/rack/methodoverride.rb:22:in `call'
  rack (1.6.5) lib/rack/runtime.rb:18:in `call'
  activesupport (4.2.7.1) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  rack (1.6.5) lib/rack/lock.rb:17:in `call'
  actionpack (4.2.7.1) lib/action_dispatch/middleware/static.rb:120:in `call'
  rack (1.6.5) lib/rack/sendfile.rb:113:in `call'
  railties (4.2.7.1) lib/rails/engine.rb:518:in `call'
  railties (4.2.7.1) lib/rails/application.rb:165:in `call'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:155:in `handle'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:109:in `rescue in block (2 levels) in start'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:106:in `block (2 levels) in start'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:96:in `each'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:96:in `block in start'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:76:in `loop'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:76:in `start'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/lib/nack/server.rb:12:in `run'
  /Users/joel/Library/Application Support/Pow/Versions/0.5.0/node_modules/nack/bin/nack_worker:4:in `<main>'`

**Comment by jgrannas** [comment]
Update. I can resolve the issue by reverting to previous... 
`gem 'carrierwave', '~> 0.11.2'`

**Comment by doiyuki** [comment]
This bug occur on uploading  jpg file. I had no error on uploading png file.

**Comment by abepetrillo** [comment]
Reproduced n v1.1.0

**Comment by olegsobchuk** [comment]
try `mount_uploaders :images, CarUploader` instead of `mount_uploader :images, CarUploader`. `mount_uploaders` in plural

**Comment by NaiveCAI** [comment]
@olegsobchuk Really! Save my several days 😂 How did it happen. Did API change?  It's like magic.

**Comment by taufnrsyd** [comment]
Ah, my bad. I forgot to add `@model.save` afterwards. The above works fine. Sorry @olegsobchuk!

**Comment by hagarabdelwaha** [comment]
write in your model(mount_uploaders) in order to upload multiple file , if you want to upload one file write(mount_uploader)

**Comment by znz** [comment]
I have the same problem.

Here is minimal reproduce example:

```ruby
require 'bundler/inline'
gemfile do
  source 'https://rubygems.org'
  gem 'rails', '5.2.2'
  gem 'carrierwave', '1.3.1'
end
require 'tempfile'
CarrierWave::Uploader::Base.new.cache!(ActionDispatch::Http::UploadedFile.new(tempfile: Tempfile.new))
```

In https://github.com/carrierwaveuploader/carrierwave/commit/7d8301c26beb572a6a398788797afd2a8abb7801 , `cache_path` allows `original_filename` is nil, but `workfile_path` doesn't.
@mshibuya Is it intentional?

**Comment by mshibuya** [comment]
What @hagarabdelwaha says is everything: https://github.com/carrierwaveuploader/carrierwave/issues/2107#issuecomment-464458775

@znz Yours is another thing, please open an issue if that behavior affects your usecase.

**Comment by etewiah** [comment]
Just a note for anyone who comes across this later.  The issue that causes this is trying to pass in an array of files so of course changing mount_uploader to mount_uploaders fixes that. The other solution is to loop through the files and pass each one to a new instance of the model.

**Comment by sime** [comment]
I upgraded an old project from `0.10.0` to `1.3.2` and got stuck on the same error message. I'm uploading a single file, so `mount_uploaders` doesn't do the trick for me.

In my `process` callback, I had a `rescue` when the file failed to be processed. Previously within the rescue, I attempted to `remove!` the file. Commenting out that line allowed my tests to past.

Haven't found out yet how to 'nicely' `remove!` as unprocessable file.
